### ISSUE TITLE
Can we, please, have Pandoc-style non-breaking spaces?

### ISSUE THREAD

**Issue by stroobandt** [executablebooks/MyST-Parser#805]
### Describe the feature you'd like to request

For MyST to enter into my academic writing workflow, I really would like to see a MyST Markdown extension for quick-to-type non-breaking spaces.

Rationale: Scientific and technical documents are brimming with in-text values of quantities and their respective units. According to [The International System of Units](https://www.bipm.org/documents/20126/41483022/SI-Brochure-9-EN.pdf/) §5.4.3:

> The numerical value always precedes the unit and a space is always used to separate the
unit from the number. Thus the value of the quantity is the product of the number and the
unit. The space between the number and the unit is regarded as a multiplication sign (just as
a space between units implies multiplication).

Hence, line-breaking/word-wrapping between the value of a quantity and its unit is not considered best practice in typography.

A similar argument holds for the thin space between digit grouping, e.g. between thousands in numbers with more than 4 digits. See for example _Astronomy & Astrophysics_ [Typography: general typing rules](https://www.aanda.org/for-authors/latex-issues/typography).

### Describe the solution you'd like

I would love to see a Markdown extension similar to the one [available in Pandoc](https://pandoc.org/MANUAL.html#extension-all_symbols_escapable):

> A backslash-escaped space is parsed as a nonbreaking space. In TeX output, it will appear as ~. In HTML and XML output, it will appear as a literal unicode nonbreaking space character (note that it will thus actually look “invisible” in the generated HTML source […]

Hence, typing `\ ` in MyST Markdown would be interpreted as a non-breaking space `A0` in UTF-8, `&nbsp;` in HTML and `~` in LaTeX.
Similarly, typing `\,` in MyST Markdown should be interpreted as a narrow non-breaking space `202F` in UTF-8, `&nnbsp;` in HTML and `\,` in LaTeX.

### Describe alternatives you've considered

I am currently helping myself in MyST Markdown with `&nbsp;`, but typing six characters for one simple non-breaking space is extremely cumbersome. Moreover, it renders the MyST Markdown source less readable, whereas Markdown was intended for readability.

**Comment by agoose77** [comment]
Hi @stroobandt!

Over at https://github.com/executablebooks/mystmd there's work ongoing to support the siunitx syntax via a `{si}` unit role, e.g. `{si}`5 <kg.m/s^2>`. Would this meet your needs? 

If so, we ought to consider how to bring this to MyST Parser somehow.

**Comment by stroobandt** [comment]
@agoose77 That is certainly a very welcome development. However, there are more instances besides units where quick-to-type non-breaking spaces are desired.

For example, inside a narrow table cell, a non-breaking space can be useful to prevent an unfortunate word wrap. Another example, might be the space between proper names and I am undoubtedly still missing many more use cases where non-breaking spaces are useful.

In conclusion, I still would like to push for a `\ ` MyST Markdown extension.

**Comment by dbitouze** [comment]
> I am currently helping myself in MyST Markdown with `&nbsp;`, but typing six characters for one simple non-breaking space is extremely cumbersome. Moreover, it renders the MyST Markdown source less readable, whereas Markdown was intended for readability.

AFAICS, at least for an HTML output, it is enough to directly type in the MyST Markdown source the `U202F` space (on my French AZERTY keyboard on GNU/Linux: <kbd>AltGr</kbd>+<kbd>Shift</kbd>+<kbd>SPC</kbd>).

**Comment by stroobandt** [comment]
> AFAICS, at least for an HTML output, it is enough to directly type in the MyST Markdown source the `U202F` space

That is not entirely correct. `U202F` corresponds to a _narrow,_ non-breaking _thin_ space for use with units, whereas `UA0` yields the standard width non-breaking space. (Incidentally, I happen to have a [web page dedicated to this very topic](https://hamwaves.com/utf-8/en/index.html#spacing).)

However, this little mistake perfectly illustrates the issue. _The whole idea behind Markdown is for it to be human readable,_ allowing one to avoid such mistakes when editing or proof-reading a Markdown source document.

> PHILOSOPHY
Markdown is intended to be as easy-to-read and easy-to-write as is feasible.
_Source:_ <https://daringfireball.net/projects/markdown/syntax#philosophy>

When I type <kbd>Ctrl</kbd><kbd>Shift</kbd><kbd>U</kbd>`A0`, or `202F`, my text editor (Vim) will render in both cases an ordinary looking space, indistinguishable from a normal breaking space and hence not positively identifiable as any particular non-breaking space. This is not the case with Pandoc-style `\ `, respectively `\,`. Moreover, both these codings are much shorter to type and there is no need to memorise any awkward UTF-8 codes. 

> (on my French AZERTY keyboard on GNU/Linux: AltGr+Shift+SPC).

Again, the same issue. I type on the [US International](https://upload.wikimedia.org/wikipedia/commons/2/22/KB_US-International.svg) ANSI keyboard layout on GNU/Linux, which implies that I also have a right side <kbd>AltGr</kbd> key apart from the left <kbd>Alt</kbd>. Nonetheless, <kbd>Shift</kbd><kbd>AltGr</kbd><kbd>Space</kbd> produces here an ordinary breakable space and nothing is telling me so. I actually had to run the `cat -v` command on the file to see what was really happening and to learn that, with my specific keyboard layout, it is actually <kbd>Shift</kbd><kbd>Alt</kbd><kbd>Space</kbd>.

# Conclusion
Above examples clearly demonstrate the need for a concise, human readable and positively identifiable Markdown extension for coding the standard width non-breaking space as well as the thin non-breaking space. To avoid disparity, Pandoc-style `\ ` and `\,` are proposed.

**Comment by dbitouze** [comment]
> > AFAICS, at least for an HTML output, it is enough to directly type in the MyST Markdown source the `U202F` space
> 
> That is not entirely correct. `U202F` corresponds to a _narrow,_ non-breaking _thin_ space for use with units,

I don't see anything about _narrow_ or _thin_ space in [The International System of Units (SI) brochure](https://www.bipm.org/documents/20126/41483022/SI-Brochure-9.pdf): only ordinary spaces are mentioned (see e.g. § “5.4.3 Formatting the value of a quantity”, p. 149 (151 of the PDF).

> whereas `UA0` yields the standard width non-breaking space.

You're right, my bad: AFAICS, it's `UA0`.

> However, this little mistake perfectly illustrates the issue. _The whole idea behind Markdown is for it to be human readable,_ allowing one to avoid such mistakes when editing or proof-reading a Markdown source document.
> 
> > PHILOSOPHY
> > Markdown is intended to be as easy-to-read and easy-to-write as is feasible.
> > _Source:_ https://daringfireball.net/projects/markdown/syntax#philosophy

Okay.
 
> When I type CtrlShiftU`A0`, or `202F`, my text editor (Vim) will render in both cases an ordinary looking space, indistinguishable from a normal breaking space and hence not positively identifiable as any particular non-breaking space.

In my better editor (`Emacs`) :wink:, such a space is rendered as underlined, which is definitively distinguishable from an ordinary space.

> This is not the case with Pandoc-style `\ `, respectively `\,`. Moreover, both these codings are much shorter to type

Not much shorter than <kbd>AltGr</kbd>+<kbd>Shift</kbd>+<kbd>SPC</kbd> (especially on a French AZERTY keyboard).

> and there is no need to memorise any awkward UTF-8 codes.

No need to memorize any awkward UTF-8 codes with <kbd>AltGr</kbd>+<kbd>Shift</kbd>+<kbd>SPC</kbd>.
